blogposts/2021/postgres-version-update.md

```diff
@@ -0,0 +1,139 @@
+
+The **obvious approach** would be to migrate all rows in a table associated with the same index at once. Unfortunately, indexes can be very massive. Some enterprise customers are regularly uploading 13GB index files. Thus, our migration batch size is likely going to be too small to guarantee all rows can be migrated, and increasing the batch size will add additional memory and compute constraints on the migrating container.
+
+The **approach we chose** tracks the minimum and maximum row versions for each index in a separate, _much_ smaller table.
+
+*lsif\_data\_definitions*:
+
+| index_id | scheme | identifier | data | version | 
+| -------- | ------ | ---------- | ---- | ------- | 
+| 1        | foo    | bar        | .... | 1       | 
+| 1        | bar    | baz        | .... | 2       | 
+| 2        | baz    | quux       | .... | 2       | 
+| 2        | quux   | bonk       | .... | 2       |
+| 3        | bonk   | honk       | .... | 1       |
+
+*lsif\_data\_definitions\_schema\_versions*:
+
+| index_id | min_version | max_version |
+| -------- | ----------- | ----------- |
+| 1        | 1           | 2           |
+| 2        | 2           | 2           |
+| 3        | 1           | 1           |
+
+This second table allows us to efficiently determine which indexes have been migrated and which still need attention. For example, if we are currently migrating from schema version 1 to 2, then the table above shows index 1 is partially migrated, index 2 is fully migrated, and index 3 has yet to be touched.
+
+It is an efficient operation to count both the total number of indexes as well as the number of indexes that are above or below a particular schema version.
+
+The schema version bounds can be kept in sync with the source table trivially by using triggers on inserts, updates, and deletes: if we insert a new row with a new lowest or new highest version or if we update/delete a row with the last remaining lowest or highest version, then we update the associated minimum or maximum version for that index.
+
+### The catch
+
+After processing an LSIF index file, we insert data in Postgres utilizing very large bulk insert operations. During this process, we insert tens to hundreds of thousands of rows over multiple tables for each index - but we don't make tens of thousands of requests.
+
+For each insertion query, we supply as many rows as possible - up to `65535 / (# cols)` rows at a time. This reduces the total number of round trips to the database.
+
+As of this post, our [Cloud](https://sourcegraph.com/search) environment has a single index with ~390k rows in the definitions table alone, and each index has over 12k rows in this table on average. 12k rows can be inserted with a single statement, and 390k rows can be inserted over 30 statements.
```

Comment:
```suggestion
As of this post, our [Cloud](https://sourcegraph.com/search) environment has a single index with ~390k rows in the definitions table alone, and each index has over 12k rows in this table on average; 12k rows can be inserted with a single statement, and 390k rows can be inserted over 30 statements.
```

Reply:
@efritz I think it would be clearer if you combined this paragraph with the next two to signal it's all part of the same thought.